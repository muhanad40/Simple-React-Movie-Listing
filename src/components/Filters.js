import React, { Component, Fragment } from 'react';

import Slider from 'rc-slider';

export default class Filters extends Component {
  constructor() {
    super();
    this.onFilterChange = this.onFilterChange.bind(this);
    this.onRatingSliderChange = this.onRatingSliderChange.bind(this);
    this.state = {
      selectedGenreIds: []
    };
  }

  onRatingSliderChange(minRating) {
    this.props.filterByRating(minRating);
  }

  onFilterChange(e) {
    const isChecked = e.target.checked;
    let selectedGenreIds = this.state.selectedGenreIds;

    if (isChecked) {
      selectedGenreIds = [
        ...this.state.selectedGenreIds,
        parseInt(e.target.value, 10)
      ];
    } else if (!isChecked) {
      selectedGenreIds = this.state.selectedGenreIds.filter(genreId => genreId !== parseInt(e.target.value, 10));
    }

    this.setState({
      selectedGenreIds
    }, () => {
      this.props.filterByGenre(selectedGenreIds);

      // If last filter is unchecked, show all movies (aka clone movies for filter)
      if (selectedGenreIds.length === 0) {
        this.props.cloneMoviesForFiltering();
      }
    });
  }

  render() {
    return (
      <Fragment>
        <div className="filters-device-overlay"></div>

        <aside className="filters">
          <div className="filters__device-close-btn">
            <svg width="17" height="17" xmlns="http://www.w3.org/2000/svg">
              <path d="M15.783 13.635l-5.518-5.518 5.518-5.52a1.518 1.518 0 0 0 0-2.15 1.517 1.517 0 0 0-2.15 0L8.115 5.966 2.597.446a1.517 1.517 0 0 0-2.15 0 1.518 1.518 0 0 0 0 2.152l5.518 5.519-5.518 5.518a1.518 1.518 0 0 0 0 2.151 1.517 1.517 0 0 0 2.15 0l5.518-5.519 5.518 5.52a1.517 1.517 0 0 0 2.15 0 1.525 1.525 0 0 0 0-2.152z" fill="currentColor" fillRule="nonzero"/>
            </svg>
          </div>

          <div className="filters__item">
            <h2 className="filters__title">Genres</h2>

            <ul className="genres">
              {this.props.genres.map(genre => (
                <li key={ genre.id + genre.name }>
                  <input onChange={ this.onFilterChange } value={ genre.id } type="checkbox" id={ `genre-${ genre.id }` } />

                  <label htmlFor={ `genre-${ genre.id }` }>{ genre.name }</label>
                </li>
              ))}
            </ul>
          </div>

          <div className="filters__item">
            <h2 className="filters__title">Rating</h2>

            <Slider
              className="rating-slider"
              onChange={ this.onRatingSliderChange }
              defaultValue={ this.props.minRating }
              step={ 0.5 }
              min={ 0 }
              max={ 10 }
              handle={ (props) => {
                return (
                  <div
                    className="rating-slider__handle"
                    style={{
                      left: `${props.offset}%`
                    }}
                  >
                    <span>{ props.value }</span>
                  </div>
                );
              }}
            />
          </div>
        </aside>
      </Fragment>
    );
  }
}