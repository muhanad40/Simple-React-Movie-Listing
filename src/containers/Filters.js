import React from 'react';
import { connect } from "react-redux";

import * as actions from '../actions';
import Filters from '../components/Filters';

const mapStateToProps = (state) => {
  return {
    genres: state.data.availableGenres
  };
};

export default connect(mapStateToProps, actions)(Filters);