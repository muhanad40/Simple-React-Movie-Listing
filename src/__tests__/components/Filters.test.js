import Filters from '../../components/Filters';
import mockGenres from '../../__mocks__/genres.json';
import { setupCreator } from '../test-utils';

const initProps = {
  cloneMoviesForFiltering: jest.fn(),
  filterByGenre: jest.fn(),
  genres: mockGenres.genres,
  filterByRating: jest.fn()
};
const setup = setupCreator(Filters, initProps);

describe('Filters component', () => {
  let wrapper, instance;

  beforeAll(() => {
    ({ wrapper, instance } = setup());
  });

  it('should render correctly', () => {
    expect(instance).toMatchSnapshot();
  });

  it('should add selected filters to state', () => {
    const genresCheckboxes = wrapper.find('.genres input');

    genresCheckboxes.first().simulate('change', {
      target: {
        checked: true,
        value: mockGenres.genres[0].id
      }
    });

    expect(wrapper.state().selectedGenreIds).toEqual([mockGenres.genres[0].id]);
  });

  it('should remove unchecked filters from state', () => {
    wrapper.setState({
      selectedGenreIds: [
        mockGenres.genres[0].id,
        mockGenres.genres[1].id
      ]
    });
    const genresCheckboxes = wrapper.find('.genres input');

    genresCheckboxes.first().simulate('change', {
      target: {
        checked: false,
        value: mockGenres.genres[0].id
      }
    });

    expect(wrapper.state().selectedGenreIds).toEqual([ mockGenres.genres[1].id ]);
  });

  it('should add selected filters to existing state and call `filterByGenre` prop with genres IDs passed as arg', () => {
    wrapper.setState({
      selectedGenreIds: [1, 4]
    });
    const genresCheckboxes = wrapper.find('.genres input');
    const expectedGenreIds = [1, 4, mockGenres.genres[0].id];

    genresCheckboxes.first().simulate('change', {
      target: {
        checked: true,
        value: mockGenres.genres[0].id
      }
    });

    expect(wrapper.state().selectedGenreIds).toEqual(expectedGenreIds);
    expect(initProps.filterByGenre).toHaveBeenCalledWith(expectedGenreIds);
  });

  it('should clone movies for filtering once last genre is unchecked', () => {
    wrapper.setState({
      selectedGenreIds: [
        mockGenres.genres[0].id
      ]
    });
    const genresCheckboxes = wrapper.find('.genres input');

    genresCheckboxes.first().simulate('change', {
      target: {
        checked: false,
        value: mockGenres.genres[0].id
      }
    });

    expect(initProps.cloneMoviesForFiltering).toHaveBeenCalledTimes(1);
  });

  it('should pass change callback to Slider', () => {
    expect(wrapper.find('.rating-slider').props().onChange).toEqual(instance.onRatingSliderChange);
  });

  it('should call `filterByRating` prop when slider callback is called', () => {
    instance.onRatingSliderChange(4);

    expect(initProps.filterByRating).toHaveBeenCalledWith(4);
  });
});