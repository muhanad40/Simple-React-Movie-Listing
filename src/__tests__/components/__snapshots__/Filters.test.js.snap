// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters component should render correctly 1`] = `
Filters {
  "context": Object {},
  "onFilterChange": [Function],
  "onRatingSliderChange": [Function],
  "props": Object {
    "cloneMoviesForFiltering": [MockFunction],
    "filterByGenre": [MockFunction],
    "filterByRating": [MockFunction],
    "genres": Array [
      Object {
        "id": 28,
        "name": "Action",
      },
      Object {
        "id": 29,
        "name": "Drama",
      },
      Object {
        "id": 30,
        "name": "Adventure",
      },
    ],
  },
  "refs": Object {},
  "setState": [Function],
  "state": Object {
    "selectedGenreIds": Array [],
  },
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_element": <Filters
        cloneMoviesForFiltering={[MockFunction]}
        filterByGenre={[MockFunction]}
        filterByRating={[MockFunction]}
        genres={
          Array [
            Object {
              "id": 28,
              "name": "Action",
            },
            Object {
              "id": 29,
              "name": "Drama",
            },
            Object {
              "id": 30,
              "name": "Adventure",
            },
          ]
        }
      />,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_newState": null,
      "_rendered": <React.Fragment>
        <div
          className="filters-device-overlay"
        />
        <aside
          className="filters"
        >
          <div
            className="filters__device-close-btn"
          >
            <svg
              height="17"
              width="17"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.783 13.635l-5.518-5.518 5.518-5.52a1.518 1.518 0 0 0 0-2.15 1.517 1.517 0 0 0-2.15 0L8.115 5.966 2.597.446a1.517 1.517 0 0 0-2.15 0 1.518 1.518 0 0 0 0 2.152l5.518 5.519-5.518 5.518a1.518 1.518 0 0 0 0 2.151 1.517 1.517 0 0 0 2.15 0l5.518-5.519 5.518 5.52a1.517 1.517 0 0 0 2.15 0 1.525 1.525 0 0 0 0-2.152z"
                fill="currentColor"
                fillRule="nonzero"
              />
            </svg>
          </div>
          <div
            className="filters__item"
          >
            <h2
              className="filters__title"
            >
              Genres
            </h2>
            <ul
              className="genres"
            >
              <li>
                <input
                  id="genre-28"
                  onChange={[Function]}
                  type="checkbox"
                  value={28}
                />
                <label
                  htmlFor="genre-28"
                >
                  Action
                </label>
              </li>
              <li>
                <input
                  id="genre-29"
                  onChange={[Function]}
                  type="checkbox"
                  value={29}
                />
                <label
                  htmlFor="genre-29"
                >
                  Drama
                </label>
              </li>
              <li>
                <input
                  id="genre-30"
                  onChange={[Function]}
                  type="checkbox"
                  value={30}
                />
                <label
                  htmlFor="genre-30"
                >
                  Adventure
                </label>
              </li>
            </ul>
          </div>
          <div
            className="filters__item"
          >
            <h2
              className="filters__title"
            >
              Rating
            </h2>
            <ComponentEnhancer(undefined)
              activeDotStyle={Object {}}
              className="rating-slider"
              defaultValue={3}
              disabled={false}
              dotStyle={Object {}}
              dots={false}
              handle={[Function]}
              handleStyle={
                Array [
                  Object {},
                ]
              }
              included={true}
              marks={Object {}}
              max={10}
              min={0}
              onAfterChange={[Function]}
              onBeforeChange={[Function]}
              onChange={[Function]}
              prefixCls="rc-slider"
              railStyle={Object {}}
              step={0.5}
              trackStyle={
                Array [
                  Object {},
                ]
              }
              vertical={false}
            />
          </div>
        </aside>
      </React.Fragment>,
      "_rendering": false,
      "_updater": [Circular],
    },
  },
  Symbol(enzyme.__setState__): [Function],
}
`;
